q2_feature_table/tests/test_subsample.py

```diff
@@ -0,0 +1,56 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2017, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from unittest import TestCase, main
+
+import numpy as np
+import numpy.testing as npt
+from biom.table import Table
+
+from q2_feature_table import subsample
+
+
+class SubsampleTests(TestCase):
+
+    def test_subsample_samples(self):
+        t = Table(np.array([[0, 1, 3], [1, 1, 2]]),
+                  ['O1', 'O2'],
+                  ['S1', 'S2', 'S3'])
+        a = subsample(t, 2, 'sample')
+        self.assertEqual(a.shape, (2, 2))
+
+        sample_ids = frozenset(a.ids(axis='sample'))
+        self.assertIn(sample_ids, set([frozenset(['S1', 'S2']),
+                                       frozenset(['S1', 'S3']),
+                                       frozenset(['S2', 'S3'])]))
+        self.assertEqual(set(a.ids(axis='observation')), set(['O1', 'O2']))
+
+        for i in a.ids(axis='sample'):
+            npt.assert_equal(t.data(i, axis='sample'),
+                             a.data(i, axis='sample'))
+
+    def test_subsample_features(self):
+        t = Table(np.array([[0, 1, 3], [1, 1, 2]]).T,
+                  ['O1', 'O2', 'O3'],
+                  ['S1', 'S2'])
+        a = subsample(t, 2, 'feature')
+        self.assertEqual(a.shape, (2, 2))
+
+        sample_ids = frozenset(a.ids(axis='observation'))
+        self.assertIn(sample_ids, set([frozenset(['O1', 'O2']),
+                                       frozenset(['O1', 'O3']),
+                                       frozenset(['O2', 'O3'])]))
+        self.assertEqual(set(a.ids(axis='sample')), set(['S1', 'S2']))
+
+        for i in a.ids(axis='observation'):
+            npt.assert_equal(t.data(i, axis='observation'),
+                             a.data(i, axis='observation'))
+
```

Comment:
I think there's a few more edge cases to test:

```
   f1
s1 0
s2 1
```
where `subsample_depth=1`, `axis='sample'` and `s1` is picked. Then the same situation for the other axis.
This should probably error since there wouldn't be any features/samples left in the table.

Then if the axis you are sampling has no ids, then the `sampling_depth <= ids` should prevent that from succeeding, but it's still possible for the opposing axis to be null. If the above edge cases error, then I would probably expect this situation to error as well, because the end result is an empty table along an axis.

Reply:
:+1: for erroring whenever an empty table would be produced (we're trying to disallow empty feature tables in qiime2, there's an open issue somewhere to make that a reality someday)

Reply:
Will QIIME allow for features or observations to be empty as in the above example? Ideally they're removed the the table since they're null.

`Table.subsample` returns empty tables without raising an exception; this isn't a problem for the underlying API. We can, however, test and raise an error within the plugin method. What error do you suggest?

Reply:
@wasade, what about something along the lines of 
`Subsampling has removed all features from the table (0-sum features are automatically removed)` and then similar for `samples` too.

Reply:
Also, since it looks like you can have a depth greater than the number of elements in an axis, we should probably guard the input for null axes (and test that too). Alternatively, `subsample` could choose to error if the `subsample_depth` exceeds the number of elements in an axis.

I'm fine with either approach.

Reply:
Cool cool, I think those last changes resolve it?